plugins/modules/foreman_hostgroup.py

```diff
@@ -0,0 +1,216 @@
+    description: Architecture name
+    required: False
+    default: None
+  domain:
+    description: Domain name
+    required: False
+    default: None
+  environment:
+    description: Environment name
+    required: False
+    default: None
+  medium:
+    description: Medium name
+    required: False
+    default: None
+  operatingsystem:
+    description: Operatingsystem name
+    required: False
+    default: None
+  parameters:
+    description: List of parameters and values
+    required: false
+    default: None
+  partition_table:
+    description: Partition table name
+    required: False
+    default: None
+  pxe_loader:
+    description: PXE Loader
+    required: False
+    default: None
+  realm:
+    description: Realm name
+    required: false
+    default: None
+  root_pass:
+    description: root password
+    required: false
+    default: None
+  force_update:
+    description: forces update of hostgroup, usefull when in need of password update
+    required: false
+    default: false
+  smart_proxy:
+    description: Smart Proxy name
+    required: False
+    default: None
+  subnet:
+    description: Subnet name
+    required: False
+    default: None
+  server_url:
+    description: foreman url
+    required: true
+  username:
+    description: foreman username
+    required: true
+  password:
+    description: foreman user password
+    required: true
+  validate_certs:
+    aliases: [ verify_ssl ]
+    description: verify ssl connection when communicating with foreman
+    default: true
+    type: bool
+  state:
+    description: domain presence
+    default: present
+    choices: ["present", "absent"]
+'''
+
+EXAMPLES = '''
+- name: "Create a Hostgroup"
+  foreman_hostgroup:
+    name: "new_hostgroup"
+    architecture: "architecture_name"
+    operatingsystem: "operatingsystem_name"
+    medium: "media_name"
+    ptable: "Partition_table_name"
+    server_url: "https://foreman.example.com"
+    username: "admin"
+    password: "secret"
+    validate_certs: False
+    state: present
+
+- name: "Update a Hostgroup"
+  foreman_hostgroup:
+    name: "new_hostgroup"
+    architecture: "updated_architecture_name"
+    operatingsystem: "updated_operatingsystem_name"
+    medium: "updated_media_name"
+    ptable: "updated_Partition_table_name"
+    root_pass: "password"
+    server_url: "https://foreman.example.com"
+    username: "admin"
+    password: "secret"
+    validate_certs: False
+    state: present
+
+- name: "Delete a Hostgroup"
+  foreman_hostgroup:
+    name: "new_hostgroup"
+    server_url: "https://foreman.example.com"
+    username: "admin"
+    password: "secret"
+    validate_certs: False
+    state: absent
+'''
+
+RETURN = ''' # '''
+
+from ansible.module_utils.foreman_helper import ForemanEntityApypieAnsibleModule
+
+
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'description': 'description',
+    'architecture_id': 'architecture_id',
```

Comment:
historically (hah), we've not been using the `_id` notation in the module itself, so the map would contain
`'architecture': 'architecture_id',` and in the assignment below you'd have `entity_dict['architecture'] = architecture['id']` and the transformation to `architecture_id` would happen in the helpers.

I wouldn't say I'm married to that concept, but I think it's cleaner as otherwise `entity_dict` will contain both, `architecture` (with the string provided by the user) and `architecture_id` (the id you got from the API) which might look confusing.

Reply:
If I map like you suggested there is error while updating the hostgroup. After sanitizing in update part of ensure_resource function it is not taking any parameters that are to be updated. I guess it is unable to map those parameters from entity to entity_dict.

Reply:
I have updated the entity_dict and name_map as you suggested with slight changes as your suggested name_mapping was not working while updating the hostgroup.